framework/src/media/FileInputDataSource.cpp

```diff
@@ -73,32 +73,39 @@ bool FileInputDataSource::open()
 
 	if (!mFp) {
 		setAudioType(utils::getAudioTypeFromPath(mDataPath));
-
-		switch (getAudioType()) {
-		case AUDIO_TYPE_MP3:
-		case AUDIO_TYPE_AAC:
-		case AUDIO_TYPE_OPUS:
-			setDecoder(std::make_shared<Decoder>(getChannels(), getSampleRate()));
-			break;
-		case AUDIO_TYPE_FLAC:
-			/* To be supported */
-			break;
-		default:
-			/* Don't set any decoder for unsupported formats */
-			break;
-		}
+		unsigned int channel;
+		unsigned int sampleRate;
+		audio_format_type_t pcmFormat;
+		audio_type_t audioType;
+		audioType = getAudioType();
 
 		mFp = fopen(mDataPath.c_str(), "rb");
-		if (mFp) {
-			medvdbg("file open success\n");
-			start();
-			return true;
-		} else {
+		if (!mFp) {
 			meddbg("file open failed error : %d\n", errno);
 			return false;
 		}
-	}
 
+		if ((audioType == AUDIO_TYPE_MP3) || (audioType == AUDIO_TYPE_AAC) || (audioType == AUDIO_TYPE_OPUS)) {
+			if (!utils::header_parsing(mFp, audioType, &channel, &sampleRate, &pcmFormat)) {
+				meddbg("header parsing failed\n");
+				return false;
+			}
+			setSampleRate(sampleRate);
+			setChannels(channel);
+			setPcmFormat(pcmFormat);
+			setDecoder(std::make_shared<Decoder>(channel, sampleRate));   
+		} else if (audioType == AUDIO_TYPE_PCM) {
+			setSampleRate(44100);
```

Comment:
Please check code flow, for pcm case.
MediaPlayer create -> setDataSource(User set sample rate & format, otherwise use feault value in media player)
-> prepare(datasource.open/* overwrite again???*/  -> set_audio_streram_out) -> start

Reply:
Also when I give some codes as a comment, it is kind of pseudo code, you should use it as a reference.
Here also switch-case is more good for readability.